File: code/c/types/test-alias.c
/* program: test-alias.c */

// Basic alias - number is an alias for int
typedef int number;

// Array alias - five_numbers is an alias for an array of five integers
typedef int five_numbers[5];

// Basic, and Array, c_string and my_string are alias
typedef char *c_string, my_string[256];

// Declaration of const_c_string, a constant c string
typedef const char *const_c_string;

void test(const_c_string text, number idx, five_numbers data)
{
    printf("%s - %d \n", text, data[idx]);
}

int main()
{
    number var1 = 3;
    five_numbers arr1 = {1, 2, 3, 4, 5};
    five_numbers grid[2] = {{1, 2, 3, 4, 5}, 
                            {6, 7, 8, 9, 10} };
    my_string name = "Fred";
    
    printf("%s\n", name);
    
    test("Hello World", var1, arr1);
    test("Grid 0 - ", var1, grid[0]);
    test("Grid 1 - ", var1, grid[1]);
    
    return 0;
}

File: code/pascal/types/TestAlias.pas
program TestAlias;

type 
    // Basic alias - number is an alias for int
    Number = Integer;
    
    // Array alias - FiveNumbers is an alias for an array of five integers
    FiveNumbers = array [0..4] of Integer;
    
procedure Test(idx: Number; data: FiveNumbers);
begin
    WriteLn(' -> ', data[idx]);
end;

procedure Main();
var
    var1: Number = 3;
    arr1: FiveNumbers = (1, 2, 3, 4, 5);
    grid: array [0..1] of FiveNumbers = 
            ((1, 2, 3, 4, 5), 
             (6, 7, 8, 9, 10) );
begin
    Test(var1, arr1);
    Test(3, grid[0]);
    Test(var1, grid[1]);
end;

begin
    Main();
end.
